src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1625,34 +1629,66 @@ RouterJob(Query *originalQuery, PlannerRestrictionContext *plannerRestrictionCon
 
 	if (originalQuery->commandType == CMD_SELECT)
 	{
-		job->taskList = SingleShardSelectTaskList(originalQuery, relationShardList,
-												  placementList, shardId);
+		job->taskList = SingleShardSelectTaskList(originalQuery, job->jobId,
+												  relationShardList, placementList,
+												  shardId);
+
+		/*
+		 * Queries to reference tables, or distributed tables with multiple replica's have
+		 * their task placements reordered according to the configured
+		 * task_assignment_policy. This is only applicable to select queries as the modify
+		 * queries will be reordered to _always_ use the first-replica policy during
+		 * execution.
+		 */
+		ReorderTaskPlacementsByTaskAssignmentPolicy(job, TaskAssignmentPolicy);
```

Comment:
With replication factor = 1 and reference table usage, many users would hit this and need to change the setting to spread the load on the reference tables across the cluster. (For replication factor == 1, task assignment policies don't have any affect since there is only a single placement and we have to assign it 😄 ).

Do you think would it make sense to make this behavior default? If yes, I think we need to discuss about how to make it :)

Reply:
I don't think it should be the default because it means consecutive SELECTs may flip back-and-forth between different versions of the reference table depending on whether a commit has happened on the shard placement or not (no monotonic read consistency). Using greedy/first-replica doesn't entirely prevent that, but at least makes it quite unlikely that you'll go back to an older version since it can only happen in case of failure.

Reply:
Do you mean that if we do something like the following concurrently we're more likely to get inconsistent results on session 2 and/or session 3?

```SQL
-- session 1
INSERT INTO ref_table .. ();
```

```SQL
-- session 2
SELECT count(*) FROM ref_table;

-- another select from the same session
SELECT count(*) FROM ref_table;
```


```SQL
-- session 3
SELECT count(*) FROM ref_table;
```

Or, what you've mind is something else?

Reply:
Yes, the second count(*) could see a lower value than the first count(*) when the task assignment policy is round-robin. That's an anomaly that we normally would not expect. 

Another potential downside is a worse cache hit ratio, especially because reference tables end up being redundantly cached on every node.

Reply:
@marcocitus that should only be possible if the query comes in just between commit times to workers right ?